### PR TITLE
fix(storage): compact not work as expected with add column

### LINKED ISSUE #10020: bug: compact not work as expected with alter table
**Summary**

```
mysql> create table t(a int,b int);
Query OK, 0 rows affected (0.06 sec)

mysql> insert into t values(1,1),(2,2);
Query OK, 2 rows affected (0.03 sec)

mysql> alter table t drop column b;
Query OK, 0 rows affected (0.06 sec)

mysql> alter table t add column b int default 100;
Query OK, 0 rows affected (0.04 sec)

mysql> insert into t values(3,3),(4,4);
Query OK, 2 rows affected (0.04 sec)

mysql> select * from t;
+------+------+
| a    | b    |
+------+------+
|    1 |  100 |
|    2 |  100 |
|    3 |    3 |
|    4 |    4 |
+------+------+
4 rows in set (0.04 sec)
Read 4 rows, 32.00 B in 0.011 sec., 367.63 rows/sec., 2.87 KiB/sec.

mysql> optimize table t compact ;
Query OK, 4 rows affected (0.05 sec)
```

After compact, we get the new col_stats in segment as follow:
```
"col_stats": {
	"1": {
		"min": {
			"Number": {
				"Int32": 1
			}
		},
		"max": {
			"Number": {
				"Int32": 2
			}
		},
		"null_count": 0,
		"in_memory_size": 8,
		"distinct_of_values": 3
	},
	"2": {
		"min": {
			"Number": {
				"Int32": 3
			}
		},
		"max": {
			"Number": {
				"Int32": 4
			}
		},
		"null_count": 0,
		"in_memory_size": 8,
		"distinct_of_values": 0
	},
	"0": {
		"min": {
			"Number": {
				"Int32": 1
			}
		},
		"max": {
			"Number": {
				"Int32": 4
			}
		},
		"null_count": 0,
		"in_memory_size": 16,
		"distinct_of_values": 4
	}
}
```

We need to add processing logic to the compact.

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Fill the default val when reduce column statistics.

Fix https://github.com/datafuselabs/databend/pull/10051#issuecomment-1432331928

Closes #10020

### CODE REVIEW

File: src/query/storages/fuse/src/statistics/column_statistic.rs [hunk 1]
```diff
@@ -198,6 +201,56 @@ pub mod traverse {
         }
         Ok(())
     }
+
+    pub fn traverse_scalar_dns(
```
Comment (dantengsky) [inline src/query/storages/fuse/src/statistics/column_statistic.rs:205]: is `dns` a type of search? or just depth first?

**Review by dantengsky — APPROVED** [review]

File: src/query/storages/fuse/src/statistics/column_statistic.rs (hunk 1 above)
Comment (zhyass) [reply src/query/storages/fuse/src/statistics/column_statistic.rs:205]: > is `dns` a type of search? or just depth first?

depth first

**Comment by zhyass** [comment]
```
mysql> create table t(a int,b int);
Query OK, 0 rows affected (0.06 sec)

mysql> insert into t values(0,0),(0,0);
Query OK, 2 rows affected (0.04 sec)

mysql> alter table t drop column a;
Query OK, 0 rows affected (0.04 sec)

mysql> alter table t add column a int default 0;
Query OK, 0 rows affected (0.04 sec)

mysql> insert into t values(1,1),(3,3);
Query OK, 2 rows affected (0.04 sec)

mysql> insert into t values(10,10),(30,30);
Query OK, 2 rows affected (0.04 sec)

mysql> optimize table t all;
Query OK, 6 rows affected (0.08 sec)
```
unexpected `distinct_of_values` for column `a`:
```
{
	"2": {
		"min": {
			"Number": {
				"Int32": 0
			}
		},
		"max": {
			"Number": {
				"Int32": 30
			}
		},
		"null_count": 0,
		"in_memory_size": 24,
		"distinct_of_values": 0
	},
	"1": {
		"min": {
			"Number": {
				"Int32": 0
			}
		},
		"max": {
			"Number": {
				"Int32": 30
			}
		},
		"null_count": 0,
		"in_memory_size": 24,
		"distinct_of_values": 5
	}
}
```

**Comment by dantengsky** [comment]
> unexpected `distinct_of_values` for column `a`:

:crying_cat_face:

File: src/query/storages/fuse/src/statistics/column_statistic.rs (hunk 1 above)
Comment (lichuang) [reply src/query/storages/fuse/src/statistics/column_statistic.rs:272]: `depth first search` should named to `dfs`, not `dns`

**Comment by zhang2014** [comment]
@mergify update

**Pull request merged**